File: examples/gallery/app/view/dialog_interface.py
# coding:utf-8
from PySide6.QtCore import Qt

from qfluentwidgets import (PushButton, Dialog, MessageBox, ColorDialog, TeachingTip, TeachingTipTailPosition,
                            InfoBarIcon, Flyout, FlyoutView)
from ..common.translator import Translator
from .gallery_interface import GalleryInterface


class DialogInterface(GalleryInterface):
    """ Dialog interface """

    def __init__(self, parent=None):
        t = Translator()
        super().__init__(
            title=t.dialogs,
            subtitle='qfluentwidgets.components.dialog_box',
            parent=parent
        )
        self.setObjectName('dialogInterface')

        button = PushButton(self.tr('Show dialog'))
        button.clicked.connect(self.showDialog)
        self.addExampleCard(
            self.tr('A frameless message box'),
            button,
            'https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/PySide6/examples/dialog/demo.py'
        )

        button = PushButton(self.tr('Show dialog'))
        button.clicked.connect(self.showMessageDialog)
        self.addExampleCard(
            self.tr('A message box with mask'),
            button,
            'https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/PySide6/examples/message_dialog/demo.py'
        )

        button = PushButton(self.tr('Show dialog'))
        button.clicked.connect(self.showColorDialog)
        self.addExampleCard(
            self.tr('A color dialog'),
            button,
            'https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/PySide6/examples/color_dialog/demo.py'
        )

        self.teachingButton = PushButton(self.tr('Show teaching tip'))
        self.teachingButton.clicked.connect(self.showBottomTeachingTip)
        self.addExampleCard(
            self.tr('A teaching tip'),
            self.teachingButton,
            'https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/teaching_tip/demo.py'
        )

        self.teachingRightButton = PushButton(self.tr('Show teaching tip'))
        self.teachingRightButton.clicked.connect(self.showLeftBottomTeachingTip)
        self.addExampleCard(
            self.tr('A teaching tip with image and button'),
            self.teachingRightButton,
            'https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/teaching_tip/demo.py'
        )

        self.simpleFlyoutButton = PushButton(self.tr('Show flyout'))
        self.simpleFlyoutButton.clicked.connect(self.showSimpleFlyout)
        self.addExampleCard(
            self.tr('A simple flyout'),
            self.simpleFlyoutButton,
            'https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/flyout/demo.py'
        )

        self.complexFlyoutButton = PushButton(self.tr('Show flyout'))
        self.complexFlyoutButton.clicked.connect(self.showComplexFlyout)
        self.addExampleCard(
            self.tr('A flyout with image and button'),
            self.complexFlyoutButton,
            'https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/flyout/demo.py'
        )


    def showDialog(self):
        title = self.tr('This is a frameless message dialog')
        content = self.tr(
            "If the content of the message box is veeeeeeeeeeeeeeeeeeeeeeeeeery long, it will automatically wrap like this.")
        w = Dialog(title, content, self.window())
        if w.exec():
            print('Yes button is pressed')
        else:
            print('Cancel button is pressed')

    def showMessageDialog(self):
        title = self.tr('This is a message dialog with mask')
        content = self.tr(
            "If the content of the message box is veeeeeeeeeeeeeeeeeeeeeeeeeery long, it will automatically wrap like this.")
        w = MessageBox(title, content, self.window())
        if w.exec():
            print('Yes button is pressed')
        else:
            print('Cancel button is pressed')

    def showColorDialog(self):
        w = ColorDialog(Qt.cyan, self.tr('Choose color'), self.window())
        w.colorChanged.connect(lambda c: print(c.name()))
        w.exec()

    def showBottomTeachingTip(self):
        TeachingTip.create(
            target=self.teachingButton,
            icon=InfoBarIcon.SUCCESS,
            title='Lesson 4',
            content=self.tr("With respect, let's advance towards a new stage of the spin."),
            isClosable=True,
            tailPosition=TeachingTipTailPosition.BOTTOM,
            duration=-1,
            parent=self
        )

    def showLeftBottomTeachingTip(self):
        t = TeachingTip(
            target=self.teachingRightButton,
            icon=None,
            title='Lesson 5',
            content=self.tr("The shortest shortcut is to take a detour."),
            image=":/gallery/images/Gyro.jpg",
            isClosable=True,
            tailPosition=TeachingTipTailPosition.LEFT_BOTTOM,
            duration=3000,
            parent=self.window()
        )
        button = PushButton('Action')
        button.setFixedWidth(120)
        t.addWidget(button, align=Qt.AlignRight)
        t.show()

    def showSimpleFlyout(self):
        Flyout.create(
            icon=InfoBarIcon.SUCCESS,
            title='Lesson 3',
            content=self.tr('Believe in the spin, just keep believing!'),
            target=self.simpleFlyoutButton,
            parent=self.window()
        )

    def showComplexFlyout(self):
        view = FlyoutView(
            title=self.tr('Julius·Zeppeli'),
            content=self.tr("Where the tennis ball will land when it touches the net, no one can predict. \nIf that moment comes, I hope the 'goddess' exists. \nIn that case, I would accept it no matter which side the ball falls on."),
            image=':/gallery/images/SBR.jpg',
        )

        # add button to view
        button = PushButton('Action')
        button.setFixedWidth(120)
        view.addWidget(button, align=Qt.AlignRight)

        # adjust layout (optional)
        view.widgetLayout.insertSpacing(1, 5)
        view.widgetLayout.insertSpacing(0, 5)
        view.widgetLayout.addSpacing(5)

        # show view
        Flyout.make(view, self.complexFlyoutButton, self.window())

File: examples/FluentGallery/View/DialogInterface.cpp
#include "DialogInterface.h"
#include "../Common/Translator.h"
#include <QFluentWidgets/Components/DialogBox/ColorDialog.h>
#include <QFluentWidgets/Components/DialogBox/Dialog.h>
#include <QFluentWidgets/Components/Widgets/Flyout.h>
#include <QFluentWidgets/Components/Widgets/InfoBar.h>
#include <QFluentWidgets/Components/Widgets/Label.h>
#include <QFluentWidgets/Components/Widgets/LineEdit.h>
#include <QFluentWidgets/Components/Widgets/PushButton.h>
#include <QFluentWidgets/Components/Widgets/TeachingTip.h>

DialogInterface::DialogInterface(QWidget* parent)
    : GalleryInterface(translator.dialogs, "<QFluentWidgets/Components/DialogBox>", parent)
{
    setObjectName("dialogInterface");

    // mask message dialog
    auto messageBoxButton = new PushButton(tr("Show dialog"));
    connect(messageBoxButton, &PushButton::clicked, this, &DialogInterface::showMessageDialog);
    addExampleCard(
        tr("A message box with mask"), messageBoxButton,
        "https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/dialog_flyout/message_dialog/demo.py");

    // custom message box
    auto customBoxButton = new PushButton(tr("Show dialog"));
    connect(customBoxButton, &PushButton::clicked, this, &DialogInterface::showCustomDialog);
    addExampleCard(
        tr("A custom message box"), customBoxButton,
        "https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/dialog_flyout/custom_message_box/demo.py");

    // color box
    auto colorBoxButton = new PushButton(tr("Show dialog"));
    connect(colorBoxButton, &PushButton::clicked, this, &DialogInterface::showColorDialog);
    addExampleCard(
        tr("A color dialog"), colorBoxButton,
        "https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/dialog_flyout/color_dialog/demo.py");

    // flyout
    simpleFlyoutButton_ = new PushButton(tr("Show flyout"), this);
    connect(simpleFlyoutButton_, &PushButton::clicked, this, &DialogInterface::showSimpleFlyout);
    addExampleCard(tr("A simple flyout"), simpleFlyoutButton_,
                   "https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/dialog_flyout/flyout/demo.py");

    complexFlyoutButton_ = new PushButton(tr("Show flyout"), this);
    connect(complexFlyoutButton_, &PushButton::clicked, this, &DialogInterface::showComplexFlyout);
    addExampleCard(tr("A flyout with image and button"), complexFlyoutButton_,
                   "https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/dialog_flyout/flyout/demo.py");

    teachingButton_ = new PushButton(tr("Show teaching tip"), this);
    connect(teachingButton_, &PushButton::clicked, this, &DialogInterface::showBottomTeachingTip);
    addExampleCard(
        tr("A teaching tip"), teachingButton_,
        "https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/dialog_flyout/teaching_tip/demo.py");

    teachingRightButton_ = new PushButton(tr("Show teaching tip"), this);
    connect(teachingRightButton_, &PushButton::clicked, this, &DialogInterface::showLeftBottomTeachingTip);
    addExampleCard(
        tr("A teaching tip with image and button"), teachingRightButton_,
        "https://github.com/zhiyiYo/PyQt-Fluent-Widgets/blob/master/examples/dialog_flyout/teaching_tip/demo.py");
}

void DialogInterface::showMessageDialog()
{
    auto title = tr("This is a message dialog with mask");
    auto content = tr("If the content of the message box is veeeeeeeeeeeeeeeeeeeeeeeeeery long, it will automatically "
                      "wrap like this.");
    auto w = new MessageBox(title, content, window());
    if (w->exec()) {
        qDebug() << "Yes button is pressed";
    } else {
        qDebug() << "Cancel button is pressed";
    }
}

void DialogInterface::showCustomDialog()
{
    auto w = new CustomMessageBox(window());
    if (w->exec()) {
        qDebug() << "Yes button is pressed";
    } else {
        qDebug() << "Cancel button is pressed";
    }
}

void DialogInterface::showColorDialog()
{
    auto w = new ColorDialog(Qt::cyan, tr("Choose color"), window());
    connect(w, &ColorDialog::colorChanged, this, [](const QColor& c) { qDebug() << c; });
    w->exec();
}

void DialogInterface::showSimpleFlyout()
{
    auto view = new FlyoutView("Lesson 3", tr("Believe in the spin, just keep believing!"),
                               new InfoBarIcon(InfoBarIcon::Success), "");
    Flyout::make(view, simpleFlyoutButton_, window());
}

void DialogInterface::showComplexFlyout()
{
    auto view = new FlyoutView(
        "Julius·Zeppeli",
        tr("Where the tennis ball will land when it touches the net, no one can predict. \nIf that moment comes, I "
           "hope the 'goddess' exists. \nIn that case, I would accept it no matter which side the ball falls on."),
        nullptr, ":/gallery/images/SBR.jpg");

    // add custom widget
    auto button = new PushButton("Action");
    button->setFixedWidth(120);
    view->addWidget(button, 0, Qt::AlignRight);

    // adjust layout
    view->widgetLayout()->insertSpacing(1, 5);
    view->widgetLayout()->insertSpacing(0, 5);
    view->widgetLayout()->addSpacing(5);

    Flyout::make(view, complexFlyoutButton_, window(), FlyoutAnimationType::SlideRight);
}

void DialogInterface::showBottomTeachingTip()
{
    auto view = new TeachingTipView("Lesson 4", tr("With respect, let's advance towards a new stage of the spin."),
                                    new InfoBarIcon(InfoBarIcon::Success), "", true);
    auto tip = TeachingTip::make(view, teachingButton_, -1, TeachingTipTailPosition::Bottom, window());
    connect(view, &TeachingTipView::closed, tip, &TeachingTip::close);
}

void DialogInterface::showLeftBottomTeachingTip() {
    auto pos = TeachingTipTailPosition::LeftBottom;
    auto view = new TeachingTipView("Lesson 5", tr("The shortest shortcut is to take a detour."), nullptr, "", true, pos);

    auto button = new PushButton("Action");
    button->setFixedWidth(120);
    view->addWidget(button, 0, Qt::AlignRight);

    auto tip = TeachingTip::make(view, teachingRightButton_, 3000, pos, this);
    connect(view, &TeachingTipView::closed, tip, &TeachingTip::close);
}

CustomMessageBox::CustomMessageBox(QWidget* parent)
    : MessageBoxBase(parent), titleLabel_(new SubtitleLabel(tr("Open URL"), this)), urlLineEdit_(new LineEdit(this))
{
    urlLineEdit_->setPlaceholderText(tr("Enter the URL of a file, stream, or playlist"));
    urlLineEdit_->setClearButtonEnabled(true);

    // add widget to view layout
    viewLayout()->addWidget(titleLabel_);
    viewLayout()->addWidget(urlLineEdit_);

    // change the text of button
    setYesButtonText(tr("Open"));
    setCancelButtonText(tr("Cancel"));

    widget_->setMinimumWidth(360);
    yesButton_->setDisabled(true);
    connect(urlLineEdit_, &LineEdit::textChanged, this, &CustomMessageBox::validateUrl);
}

void CustomMessageBox::validateUrl(const QString& text)
{
    yesButton_->setEnabled(QUrl(text).isValid());
}
